### PR TITLE
Fix for overlapping ways in invalidateExitAgainstDirection

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

PR fixes invalidateExitAgainstDirection in case of overlapping ways, issue #4024 

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x]  fix the issue
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/extractor/guidance/roundabout_handler.cpp [hunk 1]
```diff
@@ -131,43 +132,59 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
-                }
-            }
+                intersection_exit = index;
             else
-            {
-                roundabout_entry_first = true;
-            }
+                intersection_entry = index;
+        }
+        else if (road == size || intersection[road].angle == intersection[index].angle)
```
Comment (MoKob) [inline src/extractor/guidance/roundabout_handler.cpp]: I am unsure whether I understand this correctly. You are choosing the last road of a series of roads with same angles that is not a roundabout?

So on `(0:true), (0:false), (0:false)`, which one would you end up choosing (angle : is_roundabout). Shouldn't you compare to the `via_eid` somehow? Or is this mainly about not crashing?

Wouldn't the order of these roads be more or less random?

I feel that for a correct fix for the overarching problem, we would have to do some tie-breaking mechanisms in the road generation to ensure that these angles are always sorted correctly.

File: src/extractor/guidance/roundabout_handler.cpp [hunk 2]
```diff
@@ -131,43 +132,59 @@ void RoundaboutHandler::invalidateExitAgainstDirection(const NodeID from_nid,
...
+        {
+            road = index;
         }
     }
 
+    OSRM_ASSERT(road != size, coordinates[from_nid]);
+    OSRM_ASSERT(intersection_entry != size, coordinates[from_nid]);
+    OSRM_ASSERT(intersection_exit != size, coordinates[from_nid]);
+    if (road == size || intersection_entry == size || intersection_exit == size)
+    { // ignore wrong intersections
```
Comment (MoKob) [inline src/extractor/guidance/roundabout_handler.cpp]: Right above you assert that all of these are unequal to size. Is this simply an additional save-guard, or am I missing something?

File: src/extractor/guidance/roundabout_handler.cpp (hunk 1 above)
Comment (oxidase) [inline src/extractor/guidance/roundabout_handler.cpp]: yes, `road` is the last one in series of non-roundabout edges it is random for more than one. 
I think tie-breaking  with the  assumption that intersection.front() corresponds to `via_eid` is better and can be useful in other places.

File: src/extractor/guidance/roundabout_handler.cpp (hunk 2 above)
Comment (oxidase) [inline src/extractor/guidance/roundabout_handler.cpp]: it is just a safe-guard, but from the above comment it is not a proper solution, and with tie-breaking it is not required

**Comment by oxidase** [comment]
@MoKob the `entering_via_edge` now is swapped with the front value. The PR change is near #4038 assertion,  but the issue does not look related to the PR

**Comment by daniel-j-h** [comment]
@oxidase @MoKob - what's the state here can this be rebased onto master and then can go in to prevent master from segfaulting during extraction?

**Comment by daniel-j-h** [comment]
Also @oxidase you tagged this for 5.7.1 but it should not be in the 5.7 branch only in master, no?

**Comment by MoKob** [comment]
@daniel-j-h personally I would like to document a tie-breaking requirement for equal angles.

The example here has two roads at `0` angle. Swapping them, as done here, assures that the first road in the intersection will represent the opposite of `via_eid`, but it also does not solve the complete problem.
The order of these roads can be important, especially when it comes to roundabout counting. If the reverse of `via_eid` is to the right of the other road, we are entering the roundabout normally. Otherwise we are maybe crossing through the roundabout.
The swap that is implemented right now switches randomly between crossing and entering.
I am fine with merging this PR as is to unblock the demo-server processing, but I would strongly prefer to ticket this fix for further improvement.

**Comment by oxidase** [comment]
@daniel-j-h it is only in master, i have removed 5.7.1 tag

@MoKob to unblock master it is better to revert 2402d60429584aaa1ade632c796c3136667311ad 19494984eb6b0f15ab05d5c11dac60937d07e7e9 and  do not merge an incomplete solution.

Ordering by angles is not enough, one possibility would be to additionally order by signed curvatures at the intersection point if angles are equal (a tangentially touching smaller circle can be inside and outside of the bigger one).

**Comment by oxidase** [comment]
@MoKob i pushed another try to fix the issue. Now order is preserved, but `intersection_view` is rotated to place `entering_via_edge` at the front of `intersection_view` and  this is the expectation now for [the assertion](https://github.com/Project-OSRM/osrm-backend/blob/5ddb0ddcbc9b4596bec29a7d6f44de6478da36a4/src/extractor/guidance/intersection_generator.cpp#L383-L385). The change requires of setting 360 degrees for entries that are moved to the back of  `intersection_view`, but it must be correct for modulo 360 arithmetic.

Anyway the ordering must be fixed later, for example,  magic roundabouts are really interesting edge cases http://map.project-osrm.org/?z=18&center=51.562870%2C-1.770682&loc=51.562625%2C-1.769931&loc=51.563429%2C-1.773047&hl=en&alt=0

**Comment by MoKob** [comment]
@oxidase I don't know if any case exists that assumes angle < 360. Did you change to 359.9? 360.0 could theoretically be asserted for elsewhere. Did you do a `grep` for 360? Could be worth a shot. Otherwise I think this is the better solution 👍

**Comment by oxidase** [comment]
@MoKob i have checked OSM planet with car.profile and it finished without assertions. The angle modified in PR is used only to compute angles differences mod 360, so using 360 degrees is ok to preserve ordering.

**Pull request merged**